### PR TITLE
[prometheus] Fix collection output buffer management when its resized

### LINKED ISSUE #5661: Prometheus exporter produces zero-byte HTTP response
### Package

OpenTelemetry.Exporter.Prometheus.AspNetCore

### Package Version

| Package Name                                  | Version       |
| --------------------------------------------- | ------------- |
| OpenTelemetry.Exporter.Prometheus.AspNetCore  | 1.9.0-alpha.1 |
| OpenTelemetry.Extensions.Hosting              | 1.9.0-alpha.1 |
| OpenTelemetry.Instrumentation.AspNetCore      | 1.8.1         |
| OpenTelemetry.Instrumentation.Http            | 1.8.1         |
| OpenTelemetry.Instrumentation.Runtime         | 1.8.1         |

### Runtime Version

net8.0

### Description

Using Prometheus exporter `1.9.0-alpha.1` I'm getting a bunch of zero-byte responses from the metrics endpoint; status code is `200` and `Content-Length` header is `0`.

This issue has been verified by another user as well:
image

It does not make a difference whether I access directly/behind nginx, via http/https, via hostname/IP-address or even with/without password configured in the application.

Here are corresponding TRACE logs (I've replaced my actual password of course 😉):
image

### Steps to Reproduce

Two possibilities:

- Copy the example source code from the sources linked below into separate project.
  - [Asp.Net startup](https://github.com/JustArchiNET/ArchiSteamFarm/blob/d25883ff1bb25d2a589551396b3a3fb3317cc8ad/ArchiSteamFarm/IPC/ArchiKestrel.cs) (plugin is loaded at lines 249 and 404)
  - [Plugin](https://github.com/JustArchiNET/ArchiSteamFarm/blob/d25883ff1bb25d2a589551396b3a3fb3317cc8ad/ArchiSteamFarm.OfficialPlugins.Monitoring/MonitoringPlugin.cs) (lines 101 to 129)
  
- Set up [ArchiSteamFarm](https://github.com/JustArchiNET/ArchiSteamFarm/) in "generic" configuration and install it's "monitoring plugin". The application release is still tracking version `1.7.0-rc.1` so you either need to manually build or grab it from the build pipeline where we originally wanted to upgrade to the newest version [over here](https://github.com/JustArchiNET/ArchiSteamFarm/actions/runs/9181369538).

Until now I was not yet able to reproduce it using only localhost but only on a remote host so keep in mind that you might need to have a phone or raspberry pi at hand.

### Expected Result

Response from prometheus metrics endpoint contains metrics.

### Actual Result

Response status code is `200` and `Content-Length` header is `0`.

### Additional Context

Thank you in advance for looking into this. If I can be of any assistance, please let me know.

### PR DESCRIPTION
**Pull request by jcin193**
Fixes #5661

## Changes

When buffer is too small to fit its initial 85kB size its resized to fit the output, however the resized buffer is not used in final output. This ultimately results in overflow exception as cursor is correct and follows resized buffer. I also changed to use the same buffer variable for WriteTargetInfo, it also can resize the buffer.

## Merge requirement checklist

* [X] [CONTRIBUTING](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/CONTRIBUTING.md) guidelines followed (license requirements, nullable enabled, static analysis, etc.)
* [X] Unit tests added/updated
* [X] Appropriate `CHANGELOG.md` files updated for non-trivial changes
* [X] Changes in public API reviewed (if applicable)

### CODE REVIEW

**Comment by Kielek** [comment]
@jcin193, could you please sign CLA? Without this your contribution cannot be accepted.

**Comment by jcin193** [comment]
Yeah, I know, unfortunately I wrote this at work, so I'll see what I can do

**Comment by jcin193** [comment]
still working on it

File: src/OpenTelemetry.Exporter.Prometheus.HttpListener/CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 ## Unreleased
 
+* Fix collection output buffer management when its resized
+  ([#5661](https://github.com/open-telemetry/opentelemetry-dotnet/issues/5661))
```
Comment (reyang) [inline src/OpenTelemetry.Exporter.Prometheus.HttpListener/CHANGELOG.md:6]:
```suggestion
  ([#5676](https://github.com/open-telemetry/opentelemetry-dotnet/pull/5676))
```

**Review by reyang — APPROVED** [review]

**Review by Kielek — APPROVED** [review]

**Comment by jcin193** [comment]
can we get these tests running?

File: test/OpenTelemetry.Exporter.Prometheus.AspNetCore.Tests/PrometheusExporterMiddlewareTests.cs [hunk 2]
```diff
@@ -288,6 +288,30 @@ public async Task PrometheusExporterMiddlewareIntegration_CanServeOpenMetricsAnd
         await host.StopAsync();
     }
 
+    [Fact]
+    public async Task PrometheusExporterMiddlewareIntegration_ALotOfMetrics()
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Exporter.Prometheus.AspNetCore.Tests/PrometheusExporterMiddlewareTests.cs:292]:
```suggestion
    public async Task PrometheusExporterMiddlewareIntegration_TestBufferSizeIncrease_WithLotOfMetrics()
```

File: test/OpenTelemetry.Exporter.Prometheus.HttpListener.Tests/PrometheusHttpListenerTests.cs [hunk 3]
```diff
@@ -144,6 +144,45 @@ public void PrometheusHttpListenerThrowsOnStart()
         listener?.Dispose();
     }
 
+    [Theory]
+    [InlineData("application/openmetrics-text")]
+    [InlineData("")]
+    public async Task PrometheusExporterHttpServerIntegration_LargePayload(string acceptHeader)
```
Comment (vishweshbankwar) [inline test/OpenTelemetry.Exporter.Prometheus.HttpListener.Tests/PrometheusHttpListenerTests.cs:150]:
```suggestion
    public async Task PrometheusExporterHttpServerIntegration_TestBufferSizeIncrease_With_LargePayload(string acceptHeader)
```

**Review by vishweshbankwar — APPROVED** [review]

**Comment by JoelDavidLang** [comment]
Thanks for this fix. 🎉 I just spent over an hour trying to find out why my metrics were broken and then finding the issue fixed by a package update, thanks to this PR. You saved me a day of frustration.

**Pull request merged**